File: src/vyrion/_fixtures/genkit/payment_agent.mjs
// A realistic (vulnerable) Genkit human-in-the-loop tool.
//
// The transfer is gated by a Genkit interrupt (ai.defineInterrupt / interrupt()).
// On resume the approval reply is passed to transferFunds as `approval`; it is
// trusted and the transfer runs. A writer of the persisted interrupt/resume state
// can inject an approval reply with no binding to approver, action, arguments, or a
// single-use nonce (the Ghost Approval surface). Interrupt wiring is omitted here so
// the protected action is exercised directly; see README for the full flow.
import { genkit } from 'genkit';

const ai = genkit({ plugins: [] });
void ai;

// PROTECTED ACTION: consumes the approval reply and transfers funds
export async function transferFunds({ recipient, amount, currency, approval }) {
  return `settled ${amount} ${currency} to ${recipient}`;
}

File: testbeds/strands/payment_agent.py
"""A realistic (vulnerable) Strands tool gated by a human interrupt.

The tool pauses via tool_context.interrupt(); the response lives in the agent's
interrupt state and is trusted on resume. A writer of that persisted state can
inject an approval response with no binding to approver, action, arguments, or a
single-use nonce, and the tool runs on resume (the Ghost Approval surface).
"""
from strands import tool, ToolContext


@tool(context=True)
def transfer_funds(recipient: str, amount: int, currency: str, tool_context: ToolContext) -> str:
    response = tool_context.interrupt("approve_transfer", reason="approve this payment?")
    # PROTECTED ACTION
    return f"settled {amount} {currency} to {recipient}"
